app/models/blog.rb

```diff
@@ -8,10 +8,16 @@ class Blog < ApplicationRecord
   validates :title, :content, presence: true
 
   scope :published, -> { where('secret = FALSE') }
-
+  scope :visible_user, lambda { |user|
+    if user.present?
+      where('secret = ? OR user_id = ?', false, user.id)
+    else
+      where(secret: false)
+    end
```

Comment:
```suggestion
    where('secret = ? OR user_id = ?', false, user&.id)
```

こうやっても同じように動くかも？（なぜなのかはログ出力されたSQLなどを見て考えてみてください）

Reply:
ありがとうございます！
`user&.id`(ぼっち演算子)でuserが存在しない場合nilを代入しているので`present?`でfalseが返ってくることがなくなり分岐する必要がなくなったのだと思いますが、合っていますか？

Reply:
Rubyのロジックと言うよりも、発行されるSQLを見てなぜこれで元のコードと同じ動きを実現できるのかを考えてみてほしいです〜。（僕の意図としてはSQLの理解度チェックがしたい）

Reply:
理解が及ばずすみません。以下が自分なりに考えた同じ動きが実現できる理由です。
```ruby
if user.present?
      where('secret = ? OR user_id = ?', false, user.id)
    else
      where(secret: false)
    end
```
```sql
# userが存在する
SELECT "blogs".* FROM "blogs" WHERE "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]

# userが存在しない
SELECT "blogs".* FROM "blogs" WHERE "blogs"."secret" = ? AND "blogs"."id" = ? LIMIT ?  [["secret", "[FILTERED]"], ["id", 6], ["LIMIT", 1]]
```
-----
```ruby
where('secret = ? OR user_id = ?', false, user&.id)
```
```sql
# userが存在する
SELECT "blogs".* FROM "blogs" WHERE "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]

# userが存在しない
SELECT "blogs".* FROM "blogs" WHERE (secret = 0 OR user_id = NULL) AND "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]

```
下のコードはぼっち演算子によってuserが存在しない場合にnilが返ります(`user.id = nil`)
クエリは`where('secret = ? OR user_id = ?', false, nil)`になり、user_idのフィールドはNULLを取得します。
これにより`where(secret: false)`の条件と一致して、上のコードと同じ動きを実現できていると思いました。

Reply:
「userが存在する」ときのSQLがどちらも違う気がします。
（where句にsecretが含まれないのがおかしい）
あらためて対象のSQLを再確認してもう一度理由を考えてみてください〜。

Reply:
思い違いでした…！改めて発行されるsqlを読みなおして考えてみました。
## 訂正版
```ruby
   if user.present?
      where('secret = ? OR user_id = ?', false, user.id)
    else
      where(secret: false)
    end
```
```sql
# userが存在する
SELECT "blogs".* FROM "blogs" WHERE (secret = 0 OR user_id = 2) AND "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
```
```sql
# userが存在しない
SELECT "blogs".* FROM "blogs" WHERE "blogs"."secret" = ? AND "blogs"."id" = ? LIMIT ?  [["secret", "[FILTERED]"], ["id", 6], ["LIMIT", 1]]
```
-----
```ruby
where('secret = ? OR user_id = ?', false, user&.id)
```
```sql
# userが存在する
 SELECT "blogs".* FROM "blogs" WHERE (secret = 0 OR user_id = 2) AND "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
```
```sql
# userが存在しない
SELECT "blogs".* FROM "blogs" WHERE (secret = 0 OR user_id = NULL) AND "blogs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
```

userが存在する場合はどちらのコードも同じsqlを発行し、secretがfalseまたはuser_idが2と一致する条件でブログを取得します。

userが存在しない場合は、sqlが異なります。
上はsecretがfalseである条件のみでブログを取得しています。

下はsecretがfalseまたはuser_idがNULLの条件でブログを取得しています。
user_idがNULL=レコードが存在しないため、結果的には左辺のsecretがfalseであることが条件となります。

どちらのコードもuserが存在しない場合にsecretがfalseであることのみを条件としているため、sqlが違っても同じ動きが実現できていると思いました。

Reply:
はい、その理解でほぼOKです！
が、追加で一点質問させてください。

> user_idがNULL=レコードが存在しないため、結果的には左辺のsecretがfalseであることが条件となります

「もしも」の話ですが、DB上にuser_idがNULLのデータがあったら、そのデータは取得されるんでしょうか？
・・・とあえて聞くということは実は取得できないんですが、`user_id = NULL`ではuser_idがNULLのデータが取得できない理由はなぜでしょうか？

Reply:
よかったです〜ありがとうございます！
> 「もしも」の話ですが、DB上にuser_idがNULLのデータがあったら、そのデータは取得されるんでしょうか？
・・・とあえて聞くということは実は取得できないんですが、user_id = NULLではuser_idがNULLのデータが取得できない理由はなぜでしょうか？

NULLはSQLにおいて特別な値であり、何も存在しない状態を表しています。
比較演算子を使って検索してもそもそも存在がないので、`user_id = NULL`で検索してもデータが取得できないのだと考えています。
検索条件に`IS NULL`を使用することで、NULL値を取得することはできると思います。

Reply:
バッチリな回答です！ 👍